src/monmove.cpp

```diff
@@ -446,10 +438,49 @@ void monster::move()
         }
     }
 
+    if( moved ) {
+        const Creature *target = g->critter_at( next, is_hallucination() );
+        // When attacking an adjacent enemy, we're direct.
+        if( target != nullptr && attitude_to( *target ) == A_HOSTILE ) {
+            // No change, stick with the plan.
+        } else {
+            // If the direct path is blocked, go around.
+            int switch_chance = 1;
+            std::vector<point> squares = squares_in_direction( pos().x, pos().y, next.x, next.y );
+            for( point &square : squares ) {
+                const tripoint candidate = { square.x, square.y, pos().z };
+                // Bail out if the move doesn't get us closer to the target.
+                if( !plans.empty() && rl_dist( candidate, plans.back() ) >=
+                    rl_dist( pos(), plans.back() ) ) {
+                    continue;
+                }
+                // Bail out if we can't move there and we can't bash.
+                if( !can_move_to( candidate ) &&
+                    !(can_bash && g->m.bash_rating( bash_estimate(), candidate ) >= 0 ) ) {
+                    continue;
+                }
+                const Creature *target = g->critter_at( next, is_hallucination() );
```

Comment:
I think it is supposed to work like this: first mon checks the exact tile it wants to move to and marks it as next, then it doesn't switch from that tile if it contains a hostile target.